7.0 account partner aged statement

mcassuto opened the pull request:
New PR of #7 in correct branch

Commit:
[ADD] new module account_partner_aged_statement

Commit:
account_partner_aged_statement_webkit: clean folder

Commit:
inherit the account module aging report while limiting to a partner

Commit:
account_partner_aged_statement_webkit - add the following :

 * the summary per period
 * the list of invoices per period
 * the invoices are also shown with the foreign currency

Commit:
account_partner_aged_statement_webkit - try to get one partner report per line

Commit:
change the query based on current partner to get partner lines

Commit:
account_partner_aged_statement_webkit: fetch the email's body from the Company's Overdue Message

mcassuto commented:
@nbessi : about your question 
"Is there a reason to not inherit account-financial-reporting / account_financial_report_webkit / report / aged_partner_balance.py instead ?"

Yes :we had a long discussion about that: as accounting will be totally reviewed, we thought it was safer to inherit from the source....

Commit:
account_partner_aged_statement - Minor fixes :
1/ aling-right for cell with amounts
2/ print the line.move_id.name rather than line_id.name which usually provides the information than line_id.ref

max3903 commented on account_partner_aged_statement_webkit/__init__.py:
```diff
@@ -0,0 +1,23 @@
+# -*- encoding: utf-8 -*-
+
+############################################################################
+#
```
Please add 

OpenERP, Open Source Management Solution
This module copyright (C) 2014 Savoir-faire Linux
(http://www.savoirfairelinux.com).

here

max3903 commented:
Can you provide the pot file ?

:-1:

Commit:
prevent extra useless second page

Commit:
clean page break fix

Commit:
style fix, copyright missing

llacroix committed:
Added translations

llacroix committed:
Update french translation

llacroix committed:
Translate header

llacroix committed:
Fix partner aged statement translation

llacroix committed:
Fix translation "Send aged statement"

llacroix committed:
Include uncessary <br> inside hidden code to prevent trailing brs

llacroix committed:
Added a missing br and fixed html formatting

Commit:
Add Supplier Aged Statement report

Commit:
Add missing french translation

Commit:
Fix supplier and partner reports so that supplier report has only payable amounts and partner has only receivable amounts. Both reports have now all amounts in the correct column with the correct sign

Commit:
Improve the report. The total and paid amounts on each invocie will appear in the same line

Commit:
Fix the order and remove unused parameter on methods

Commit:
Fix .travis.yml and fix method get_balance

Commit:
Add dependency to xfonts-75dpi in .travis.yml

Commit:
Add dependency to webkit-tools in .travis.yml

ghost commented:
@max3903 @veloutin @llacroix  Your reviews would be greatly appreciated here, this pull request needs to be merged. Thanks

max3903 commented:
:+1:

llacroix commented on account_partner_aged_statement_webkit/i18n/fr.po:
```diff
@@ -0,0 +1,237 @@
+"\n"
+"    <p style=\"border-left: 1px solid #8e0000; margin-left: 30px;\">\n"
+"       &nbsp;&nbsp;<strong>RÉFÉRENCES</strong><br />\n"
+"       &nbsp;&nbsp;Total du montant en retard: ${object.credit} ${user.company_id.currency_id.name}<br/>\n"
```
The msgid should be in english. Not in french. The msgstr isn't translated. I'm not sure the translation will be picked up by odoo as the msgid doesn't exists anywhere else.

llacroix commented on account_partner_aged_statement_webkit/partner_aged_statement_data.xml:
```diff
@@ -0,0 +1,74 @@
+    <br/>
+
+    <p style="border-left: 1px solid #8e0000; margin-left: 30px;">
+       &nbsp;&nbsp;<strong>REFERENCES</strong><br />
```
Instead of translating this whole html content, you could instead use the translating function to translate some fields only. `${_("REFERENCES")}` and 
`${_("Total overdue amount: %s %s") % (object.credit, user.company_id.currency_id.name)}`

dufresnedavid replied:
I tried this. The problem is that on rendering of the Header, Odoo doesn't recognise _().

llacroix replied:
That's kind of strange, there is an other module out there that does exactly like that.

llacroix commented on account_partner_aged_statement_webkit/partner_aged_statement_data.xml:
```diff
@@ -0,0 +1,74 @@
+        </span>
+        % if user.company_id.phone:
+            <div style="margin-top: 0px; margin-right: 0px; margin-bottom: 0px; margin-left: 0px; padding-top: 0px; padding-right: 0px; padding-bottom: 0px; padding-left: 0px; ">
+                Phone:&nbsp; ${user.company_id.phone}
```
You can also translate `${_("Phone")}` above and `${_("Web")}`. You could also check if the `nbsp` are really necessary.

llacroix commented on account_partner_aged_statement_webkit/report/partner_aged_statement.mako:
```diff
@@ -0,0 +1,174 @@
+        %else:
+            <div class="title">${_('Aged Balance')}</div>
+            <br>
+        %if get_balance(partner, company):
```
Bad indentation, this if is nested within the else above. It might be misleading.

llacroix commented on account_partner_aged_statement_webkit/report/partner_aged_statement.mako:
```diff
@@ -0,0 +1,174 @@
+            </table>
+        %endif
+        <br>
+        <br>
```
These `<br>` tags aren't closed. It's not a big deal. Having open tags prevent xml parser to parse html files. If this module was to be ported to odoo8 QWeb templates, it would be better to keep mako templates xml friendly as much as possible. Also instead of using `<br>`. It would make more sense to add a margin on block elements. Line breaks might get added multiple times if content between the blocks isn't added. For example, here if the test for `get_balance` doesn't pass, you'll have 3 line breaks instead of 1 or 2.. When those line breaks adds up at the end of the page, it might create new empty pages.

llacroix commented:
Everything seems fine except the missing translations. The other issues aren't blocking but would probably improve the quality of the code.

gdgellatly commented on account_partner_aged_statement_webkit/report/partner_aged_statement_report.py:
```diff
@@ -0,0 +1,392 @@
+        return message.split('\n')
+
+    def _get_fiscalyear(self, data):
+        now = data['fInvoicesorm']['date_from']
```
Is this right, looks like typo?

dufresnedavid replied:
What do you mean?

llacroix replied:
I guess he's talking about `fInvoicesorm` that's a strange name.

dufresnedavid replied:
This method is unused, so I removed it.

gdgellatly commented on account_partner_aged_statement_webkit/report/partner_aged_statement_report.py:
```diff
@@ -0,0 +1,392 @@
+        company_pool = pooler.get_pool(self.cr.dbname)['res.company']
+        message = company_pool.browse(
+            self.cr, self.uid, company.id, {'lang': obj.lang}).overdue_msg
+        return message.split('\n')
```
What if there is no message?

gdgellatly commented on account_partner_aged_statement_webkit/report/partner_aged_statement_report.py:
```diff
@@ -0,0 +1,392 @@
+        return self.partner_invoices_dict[partner.id]
+
+    def _lines_get_30(self, partner, company):
+        today = datetime.now()
```
Should this be set using the users timezone?  I guess it only affects people accessing servers in significantly different timezones to their own.

gdgellatly commented on account_partner_aged_statement_webkit/report/partner_aged_statement_report.py:
```diff
@@ -0,0 +1,392 @@
+        now = data['fInvoicesorm']['date_from']
+        domain = [
+            ('company_id', '=', self._company.id),
+            ('date_start', '<', now),
```
You probably want <= and >= here

gdgellatly commented on account_partner_aged_statement_webkit/report/partner_aged_statement_report.py:
```diff
@@ -0,0 +1,392 @@
+        movelines = self._get_current_invoice_lines(partner, company, today)
+        movelines = [
+            line for line in movelines
+            if stop < line['date_due'] <= start
```
check for no date_due

Commit:
Remove unused methods, add timezone to the reference date and fix translation

gdgellatly commented:
:+1:

Commit:
Fix programming error

yvaucher commented:
@mcassuto You need to either setup your local git email or add your email to github as it seems those commits are not linked to your account.

mcassuto commented:
@yvaucher CLA submitted, github email address updated too.
The situation here is this PR has been submitted 7 months while I was working at SFL; I am not there anymore and they completed this PR (which is totally ok).

ehdem committed:
[UPD]Fix bug when send aged statement

* Put default_parent_id to false if it is in context
because it could be use in mail.compose.message model
and create a bug saying xx id doesn't exit in model.
* The default_parent_id only exists when doing search with
the related company field; to fix this issue, check if the
context contains the default_parent_id and put it false

gurneyalex commented:
the empty clabot message should be fixed now. Sorry for the inconvenience.

joaoalf commented on account_partner_aged_statement_webkit/partner_aged_statement.py:
```diff
@@ -0,0 +1,76 @@
+        # to get the message_id. It must be false when doing search with
+        # related company
+        if 'default_parent_id' in context:
+            context['default_parent_id'] = False
```
@ehdem Instead of putting it to False it would be better to remove the key.

Commit:
Simplify method _get_current_invoice_lines and add unit tests

Commit:
Fix parser

Commit:
Base the report on the invoice date instead of due date

Commit:
Add missing translation

Commit:
Add invoice refunds

gurneyalex commented:
Can you rebase your PR?

Ehtaga committed:
[FIX] Fix self._partners bug, probably due to a bad merging

Ehtaga committed:
[FIX] Fix bug with invoices aged exactly of 30 or 60 days

Those invoices were in 2 different table (e.g. 30-60 and 60+ for a 60 days invoice)

max3903 commented:
Added webkit-tools repo to runbot

Commit:
Merge remote-tracking branch 'origin/7.0' into HEAD

Conflicts:
	.travis.yml
